driver/utils/unicode_conv.h

```diff
@@ -182,113 +196,84 @@ inline decltype(auto) toUTF8(const std::basic_string<CharType> & src) {
-
-    return src;
-}
-
-template <>
-inline decltype(auto) fromUTF8<char16_t>(const std::string & src) {
-    std::wstring_convert<std::codecvt_utf8<char16_t>, char16_t> convert;
-    return convert.from_bytes(src);
-}
-
-template <>
-inline decltype(auto) fromUTF8<char32_t>(const std::string & src) {
-    std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> convert;
-    return convert.from_bytes(src);
-}
-
+#if !defined(_MSC_VER) || _MSC_VER >= 1920
 template <>
-inline decltype(auto) fromUTF8<wchar_t>(const std::string & src) {
-    std::wstring_convert<std::codecvt_utf8<wchar_t>, wchar_t> convert;
-    return convert.from_bytes(src);
+inline decltype(auto) fromUTF8<char16_t>(const std::string & src, UnicodeConversionContext & context) {
+    return context.UCS2_converter_char16.from_bytes(src);
 }
 
 template <>
-inline decltype(auto) fromUTF8<signed char>(const std::string & src) {
-    const auto converted = fromUTF8<char>(src);
-    return std::basic_string<signed char>{converted.begin(), converted.end()};
+inline decltype(auto) fromUTF8<char32_t>(const std::string & src, UnicodeConversionContext & context) {
+    return context.UCS2_converter_char32.from_bytes(src);
 }
+#endif
 
 template <>
-inline decltype(auto) fromUTF8<unsigned char>(const std::string & src) {
-    const auto converted = fromUTF8<char>(src);
-    return std::basic_string<unsigned char>{converted.begin(), converted.end()};
+inline decltype(auto) fromUTF8<wchar_t>(const std::string & src, UnicodeConversionContext & context) {
+    return context.UCS2_converter_wchar.from_bytes(src);
 }
 
+#if !defined(_MSC_VER) || _MSC_VER >= 1920
 template <>
-inline decltype(auto) fromUTF8<unsigned short>(const std::string & src) {
+inline decltype(auto) fromUTF8<unsigned short>(const std::string & src, UnicodeConversionContext & context) {
     static_assert(sizeof(unsigned short) == sizeof(char16_t), "unsigned short doesn't match char16_t exactly");
-    const auto converted = fromUTF8<char16_t>(src);
+    const auto converted = fromUTF8<char16_t>(src, context);
     return std::basic_string<unsigned short>{converted.begin(), converted.end()};
 }
+#endif
 
 template <typename CharType>
-inline void fromUTF8(const std::string & src, std::basic_string<CharType> & dest, const std::locale& locale); // Leave unimplemented for general case.
-
-template <>
-inline void fromUTF8<char>(const std::string & src, std::string & dest, const std::locale& locale) {
-    dest = fromUTF8<char>(src, locale);
-}
-
-template <>
-inline void fromUTF8<signed char>(const std::string & src, std::basic_string<signed char> & dest, const std::locale& locale) {
-    dest = fromUTF8<signed char>(src, locale);
-}
-
-template <>
-inline void fromUTF8<unsigned char>(const std::string & src, std::basic_string<unsigned char> & dest, const std::locale& locale) {
-    dest = fromUTF8<unsigned char>(src, locale);
+inline decltype(auto) fromUTF8(const std::string & src) {
+    UnicodeConversionContext context;
```

Comment:
Why do you need to instantiate `UnicodeConversionContext` on the stack? IIRC, the whole purpose of this PR was to reuse `std::wstring_convert<std::codecvt_utf8<X>` instances as much as possible.

Reply:
For uses not on a critical path.